web/src/settings_components.ts

```diff
@@ -395,19 +396,26 @@ function read_select_field_data_from_form(
     $profile_field_form.find("div.choice-row").each(function (this: HTMLElement) {
         const text = $(this).find("input")[0].value;
         if (text) {
-            if (old_option_value_map.get(text) !== undefined) {
-                // Resetting the data-value in the form is
-                // important if the user removed an option string
-                // and then added it back again before saving
-                // changes.
-                $(this).attr("data-value", old_option_value_map.get(text));
-            }
             const value = $(this).attr("data-value")!;
             field_data[value] = {text, order: field_order.toString()};
             field_order += 1;
         }
     });
 
+    // Check if the user has deleted an option and added the same option again by manually
+    // filling out its value. If that's the case, we replace the new key in `field_data`
+    // with the original key from `old_option_value_map`.
+    for (const [old_choice, old_key] of old_option_value_map.entries()) {
+        // We want to find the first occurrence of the old choice, if there are multiple
+        // occurrences, we want to let the other occurrences be as is, since we rely on
+        // the API response to give us an error in case of duplicate option values.
+        const new_value_for_deleted_choice = _.findKey(field_data, ["text", old_choice]);
+
+        if (!field_data[old_key] && new_value_for_deleted_choice) {
+            field_data[old_key] = field_data[new_value_for_deleted_choice];
+            _.unset(field_data, new_value_for_deleted_choice);
```

Comment:
Can't delete dynamically because tslint won't allow it, so I have used unset here. let me know if we don't want to use `unset` here for some reason

Reply:
Noticed, we use `Map` instead in such situations, refactoring to use that ⌛